### PR TITLE
Fixed #25373 -- Added warning logging for exceptions during {% include %} tag rendering.

### PR DESCRIPTION
**Pull request by limnick**
PR for patch referenced in [ticket 25373](https://code.djangoproject.com/ticket/25373)

### CODE REVIEW

File: django/template/loader_tags.py [hunk 1]
```diff
@@ -204,9 +207,11 @@ def render(self, context):
                 return template.render(context.new(values))
             with context.push(**values):
                 return template.render(context)
-        except Exception:
+        except Exception as e:
             if context.template.engine.debug:
                 raise
+            else:
+                logger.warn('template - Uncaught Exception raised during rendering.', exc_info=e)
```
Comment (charettes) [inline django/template/loader_tags.py]: I don't think you need to use the `'template - ...'` prefix here, the log is already namespaced under `'django.template'`.

Comment (charettes) [inline django/template/loader_tags.py]: You want to either pass `exc_info=True` or `sys.exc_info()` here. `exc_info=e` only worked because it evaluated to `True`.

Comment (charettes) [inline django/template/loader_tags.py]: Ohh looks like template message are already prefixed with `'template - ...'` dismiss my first comment.

File: tests/template_tests/test_logging.py [hunk 2]
```diff
@@ -69,3 +69,37 @@ def test_log_on_variable_does_not_exist_not_silent(self):
...
+        self.logger.removeHandler(self.test_handler)
+        self.logger.level = self.original_level
+
+    def test_logs_exceptions_during_rendering_with_debug_disabled(self):
+        engine = Engine(loaders=[
+            ('django.template.loaders.locmem.Loader', {
+                'child': '{% for var in varlist %}{{ var.somemethod }}{% endfor %}',
+            }),
+        ], debug=False)
+        request = RequestFactory().get('/')
```
Comment (charettes) [inline tests/template_tests/test_logging.py]: Is there really a need to create a request in order to simulate this failure?

**Comment by limnick** [comment]
haha well i went to make the change a little too quickly. let me address all these and get another commit up

File: django/template/loader_tags.py [hunk 3]
```diff
@@ -1,4 +1,5 @@
 from collections import defaultdict
+import logging
```
Comment (charettes) [inline django/template/loader_tags.py]: Move this direct module import above `from collections` to appease isort.

File: tests/template_tests/test_logging.py (hunk 2 above)
Comment (limnick) [inline tests/template_tests/test_logging.py]: I need to give the template renderer a context with a reference to my object and was just cribbing from the existing test for the preferred way to do it. I'm open to suggestions if you have another way you think it should be done.

**Comment by limnick** [comment]
thanks for all the quick feedback!

File: django/template/loader_tags.py (hunk 1 above)
Comment (limnick) [inline django/template/loader_tags.py]: great catch on the `exc_info` I have been using it incorrectly for way too long

File: tests/template_tests/test_logging.py (hunk 2 above)
Comment (charettes) [inline tests/template_tests/test_logging.py]: I think a simple `django.template.Context` will do here.

File: tests/template_tests/test_logging.py [hunk 4]
```diff
@@ -69,3 +69,38 @@ def test_log_on_variable_does_not_exist_not_silent(self):
...
+
+        class ErrorObj():
+            def somemethod(self):
+                raise IndexError("some generic exception")
+
+        always_raises_exception = {'varlist': [ErrorObj()]}
```
Comment (charettes) [inline tests/template_tests/test_logging.py]: Unless I'm missing something you don't need to define a class and nest an instance of it in a list to reproduce your use case. Simply passing a callable that throws an exception should do.

e.g.
```python
engine = Engine(loaders=[
    ('django.template.loaders.locmem.Loader', {
        'child': '{{ raises }}',
    }),
], debug=False)

def raises():
    raise Exception

engine.from_string('{% include "child" %}').render(Context({'raises': raises}))
```

Comment (limnick) [inline tests/template_tests/test_logging.py]: I don't believe this is sufficient but will double check as I'm simplifying other aspects of the test. I believe a method call is necessary otherwise the raised exception is handled via another code path.

Comment (limnick) [inline tests/template_tests/test_logging.py]: I was wrong again. I had originally replicated the exact code path that motivated this bug, which was the django-admin detail view page. My understanding of the issue was far from complete and I had superfluous code throughout my test. Thanks again for giving such timely and detailed feedback, people like you are why Django is such a robust project!

**Comment by charettes** [comment]
Buildbot, test this please.

File: tests/template_tests/test_logging.py [hunk 5]
```diff
@@ -69,3 +69,34 @@ def test_log_on_variable_does_not_exist_not_silent(self):
...
+    def setUp(self):
+        self.test_handler = TestHandler()
+        self.logger = logging.getLogger('django.template')
+        self.original_level = self.logger.level
+        self.logger.addHandler(self.test_handler)
+        self.logger.setLevel(logging.WARN)
+
+    def tearDown(self):
+        self.logger.removeHandler(self.test_handler)
+        self.logger.level = self.original_level
```
Comment (charettes) [inline tests/template_tests/test_logging.py]: There's a lot of duplication between this test case and `VariableResolveLoggingTests.setUp` and `tearDown`.

I think it would make sense to create a common base for both of them.

e.g.
```python
class TemplateLoggingTestCase(SimpleTestCase):
    def setUp(self):
        self.test_handler = TestHandler()
        self.logger = logging.getLogger('django.template')
        self.original_level = self.logger.level
        self.logger.addHandler(self.test_handler)
        self.logger.setLevel(logging.DEBUG)

    def tearDown(self):
        self.logger.removeHandler(self.test_handler)
        self.logger.level = self.original_level
```

And assert `self.test_handler.log_record.level == logging.WARN`.

File: tests/template_tests/test_logging.py [hunk 6]
```diff
@@ -69,3 +69,34 @@ def test_log_on_variable_does_not_exist_not_silent(self):
     def test_no_log_when_variable_exists(self):
         Variable('article.section').resolve({'article': {'section': 'News'}})
         self.assertIsNone(self.test_handler.log_record)
+
+
+class ExceptionDuringRenderLoggingTests(SimpleTestCase):
```
Comment (charettes) [inline tests/template_tests/test_logging.py]: I think `IncludeNodeLoggingTests` would be a better name.

**Comment by charettes** [comment]
This is looking good! Please uncheck the ticket **Patch needs improvement flag** once you're done with the last few changes.

**Comment by berkerpeksag** [comment]
LGTM, with a couple of minor comments:

* commits need to be squashed into one commit
* commit message needs to be changed to use charettes' version
* needs a release note (see `docs/releases/1.9.txt`)
* needs a versionchanged directive in `docs/ref/templates/builtins.txt`

**Comment by limnick** [comment]
Alright those should all be taken care of. Feel free to criticize my docs changes, I tried to be as clear as possible as I expect people will go to docs when they suddenly see new exceptions being logged.

File: django/template/loader_tags.py [hunk 7]
```diff
@@ -207,6 +210,8 @@ def render(self, context):
         except Exception:
             if context.template.engine.debug:
                 raise
+            else:
+                logger.warn('template - Uncaught Exception raised during rendering.', exc_info=True)
```
Comment (timgraham) [inline django/template/loader_tags.py]: I'm not sure if the included traceback will give enough information to debug this, but it seems like a message something like "Rendering `<template name>` raised an exception, so {% include %} will render as an empty string." might be more helpful.

File: docs/ref/templates/builtins.txt [hunk 8]
```diff
@@ -673,9 +673,18 @@ if it's missing or has syntax errors), the behavior varies depending on the
 :class:`template engine's <django.template.Engine>` ``debug`` option (if not
 set, this option defaults to the value of :setting:`DEBUG`). When debug mode is
 turned on, an exception like :exc:`~django.template.TemplateDoesNotExist` or
-:exc:`~django.template.TemplateSyntaxError` will be raised; otherwise
-``{% include %}`` silences any exception that happens while rendering the
-included template and returns an empty string.
+:exc:`~django.template.TemplateSyntaxError` will be raised. When debug mode
+is turned off, ``{% include %}`` logs any exception that happens while
+rendering the included template with a level of WARNING and returns an empty
```
Comment (timgraham) [inline docs/ref/templates/builtins.txt]: add double backticks around WARNING (as done in topics/logging.txt).

File: docs/ref/templates/builtins.txt [hunk 9]
```diff
@@ -673,9 +673,18 @@ if it's missing or has syntax errors), the behavior varies depending on the
...
+string.
+
+.. versionchanged:: 1.9
+
+    Exceptions raised during template inclusion are logged using the
```
Comment (timgraham) [inline docs/ref/templates/builtins.txt]: Try to make this note "self-contained" as described in [Documenting new features](https://docs.djangoproject.com/en/dev/internals/contributing/writing-documentation/#documenting-new-features). It should be minimal, e.g. "The logging described in the previous paragraph was added."

File: docs/releases/1.9.txt [hunk 10]
```diff
@@ -528,8 +528,14 @@ Templates
 * Added a :meth:`Context.setdefault() <django.template.Context.setdefault>`
   method.
 
-* A warning will now be logged for missing context variables. These messages
-  will be logged to the :ref:`django.template <django-template-logger>` logger.
+* A DEBUG level message will now be logged for missing context variables.
```
Comment (timgraham) [inline docs/releases/1.9.txt]: Maybe we can restructure this and say something like: "The new django.template logger logs the following:" (with sub-bullets).

File: tests/template_tests/test_logging.py [hunk 11]
```diff
@@ -69,3 +73,26 @@ def test_log_on_variable_does_not_exist_not_silent(self):
...
+    loglevel = logging.WARN
+
+    def test_logs_exceptions_during_rendering_with_debug_disabled(self):
+        engine = Engine(loaders=[
+            ('django.template.loaders.locmem.Loader', {
+                'child': '{{ raisesexception }}',
+            }),
+        ], debug=False)
+
+        def errormethod():
```
Comment (timgraham) [inline tests/template_tests/test_logging.py]: error_method

File: tests/template_tests/test_logging.py [hunk 12]
```diff
@@ -69,3 +73,26 @@ def test_log_on_variable_does_not_exist_not_silent(self):
...
+            raise IndexError("some generic exception")
+
+        ctx = Context({'raisesexception': errormethod})
```
Comment (timgraham) [inline tests/template_tests/test_logging.py]: raises_exception

**Comment by timgraham** [comment]
buildbot, add to whitelist.

**Comment by limnick** [comment]
@timgraham comments much appreciated. Great call on the exception message, having a meaningful and descriptive exception is much better than sending people to the docs. These should all be cleaned up as requested.

File: django/template/loader_tags.py [hunk 13]
```diff
@@ -207,6 +210,11 @@ def render(self, context):
         except Exception:
             if context.template.engine.debug:
                 raise
+            else:
+                template_name = getattr(context, 'template_name', 'unknown')
+                logger.warn('template - Exception raised while rendering {{% include %}} '
```
Comment (timgraham) [inline django/template/loader_tags.py]: I don't understand the purpose of the "template -" prefix?

Comment (limnick) [inline django/template/loader_tags.py]: to be honest, neither do I. I thought I picked it up from a logger elsewhere in the code base but now I can't find what I thought I was referencing so I'm going to remove it.

File: django/template/loader_tags.py [hunk 14]
```diff
@@ -207,6 +210,11 @@ def render(self, context):
         except Exception:
             if context.template.engine.debug:
                 raise
+            else:
+                template_name = getattr(context, 'template_name', 'unknown')
+                logger.warn('Exception raised while rendering {{% include %}} for '
```
Comment (timgraham) [inline django/template/loader_tags.py]: Instead of interpolating the parameter in the message, logging should do that work, so this should be:
```python
logger.warn(
    'Exception raised while rendering {%% include %%} for '
    'template %s. Empty string rendered instead.',
    template_name,
    exc_info=True,
)
```
I am not sure if template name retrieval works in the 'unknown' case (seems to return "None" in the test case). Would be nice to have a test for the case when a template name does exist too.
Edits: http://dpaste.com/1XTAAT0

Comment (limnick) [inline django/template/loader_tags.py]: would you like this change applied to the other new template logger as well?

Comment (timgraham) [inline django/template/loader_tags.py]: Yes, we should probably add an actual message like you've done here and use `exc_info=True`. Please do it in a separate commit or pull request using "Refs #18773 -- ..." in the commit message (for reference, the commit is dc5b01ad05e50ccde688c73c2ed3334a956076b0). Thanks!

Comment (timgraham) [inline django/template/loader_tags.py]: Do you have time to complete these changes today? Otherwise, I'll try to take care of it before the feature freeze for 1.9.

Comment (limnick) [inline django/template/loader_tags.py]: yeah I'll get them up today.

File: django/template/base.py [hunk 15]
```diff
@@ -903,7 +903,12 @@ def _resolve_lookup(self, context):
                                 raise
         except Exception as e:
             template_name = getattr(context, 'template_name', 'unknown')
-            logger.debug('{} - {}'.format(template_name, e))
+            logger.debug(
+                'Exception raised while resolving variable %s in template %s',
+                bit,
```
Comment (limnick) [inline django/template/base.py]: I'm not thrilled with relying on the forloop above to begin executing in order for `bit` to exist in locals(). If this seems like an issue I can do something a little nastier like `locals().get('bit', 'undefined')` to get the variable name

**Comment by limnick** [comment]
docs and isort build failures here unrelated to my commit.

**Comment by limnick** [comment]
alright i'll handle these changes now. I'm hesitant to try and shift those lines into the newer commit as I don't trust my git skills enough to do that correctly.

File: django/template/base.py (hunk 15 above)
Comment (timgraham) [inline django/template/base.py]: Adding something like this before the loop could help:
```
_bit_undefined() = object() [at module level]
bit = _bit_undefined()
```

**Comment by limnick** [comment]
I've added a test case for the empty `template_name` for the import tag and fixed that issue for the existing template logger as well. The commingling of the two commits has gotten worse however, sorry about that! I'm not sure the best way to handle the `bit` being undefined in the variable resolution logger. An exception would need to occur while resolving `self.lookups` and if that's `None` then `_resolve_lookup` won't be called and if it's not iterable the user likely has a bigger problem and should get an exception from it. I'm not sure if it's an issue or not.

**Comment by timgraham** [comment]
merged in 85c52743f7b99ac6575d87080a9ffb7540acbf18 and 392f64842f678b6e29a2e5fac65a586d9c9a57ff, thanks Nick!
